godot-core/src/builtin/script.rs

```diff
@@ -0,0 +1,818 @@
+        get_property_state_func: Some(script_instance_info::get_property_state_func::<T>),
+        get_script_func: Some(script_instance_info::get_script_func::<T>),
+        get_language_func: Some(script_instance_info::get_language_func::<T>),
+        // usage is unclear.
+        get_fallback_func: None,
+        get_method_list_func: Some(script_instance_info::get_method_list_func::<T>),
+        get_property_type_func: Some(script_instance_info::get_property_type_func::<T>),
+        free_func: Some(script_instance_info::free_func::<T>),
+        free_method_list_func: Some(script_instance_info::free_method_list_func::<T>),
+        has_method_func: Some(script_instance_info::has_method_func::<T>),
+        call_func: Some(script_instance_info::call_func::<T>),
+        // deprecated by Godot.
+        notification_func: None,
+        to_string_func: Some(script_instance_info::to_string_func::<T>),
+        refcount_decremented_func: Some(script_instance_info::refcount_decremented_func::<T>),
+        refcount_incremented_func: Some(script_instance_info::refcount_incremented_func::<T>),
+        is_placeholder_func: Some(script_instance_info::is_placeholder_func::<T>),
+        // usage is unclear.
+        set_fallback_func: None,
+    };
+
+    let instance_ptr = Box::into_raw(Box::new(gd_instance));
+
+    let data = ScriptInstanceData {
+        inner: RefCell::new(rs_instance),
+        gd_instance_ptr: instance_ptr,
+    };
+
+    let data_ptr = Box::into_raw(Box::new(data));
+
+    // SAFETY: `script_instance_create` expects a `GDExtensionScriptInstanceInfoPtr` and a generic `GDExtensionScriptInstanceDataPtr` of our
+    // choice. The validity of the instance info struct is ensured by code generation.
+    unsafe {
+        sys::interface_fn!(script_instance_create)(
+            instance_ptr,
+            data_ptr as sys::GDExtensionScriptInstanceDataPtr,
+        )
+    }
+}
+
+mod script_instance_info {
+    use std::any::type_name;
+    use std::cell::{BorrowError, Ref, RefMut};
+    use std::ffi::c_void;
+    use std::mem::ManuallyDrop;
+
+    use crate::builtin::meta::GodotType;
+    use crate::builtin::{GString, StringName, Variant};
+    use crate::engine::{Object, ScriptLanguage};
+    use crate::obj::Gd;
+    use crate::private::handle_panic;
+    use crate::sys::{self, GodotFfi};
+
+    use super::{ScriptInstance, ScriptInstanceData};
+
+    fn borrow_instance_mut<'a, T: ScriptInstance>(
+        instance: *mut ScriptInstanceData<T>,
+    ) -> RefMut<'a, T> {
+        unsafe { (*instance).inner.borrow_mut() }
+    }
+
+    fn borrow_instance<'a, T: ScriptInstance>(instance: *mut ScriptInstanceData<T>) -> Ref<'a, T> {
+        unsafe { (*instance).inner.borrow() }
+    }
+
+    fn try_borrow_instance<'a, T: ScriptInstance>(
+        instance: *mut ScriptInstanceData<T>,
+    ) -> Result<Ref<'a, T>, BorrowError> {
+        unsafe { (*instance).inner.try_borrow() }
+    }
```

Comment:
References to `ScriptInstanceData<T>`? Then I have to repeat the same unsafe block across all interface functions for dereferencing the raw pointer, is that better?

Reply:
> References to `ScriptInstanceData<T>`? Then I have to repeat the same unsafe block across all interface functions for dereferencing the raw pointer, is that better?

you have to repeat the unsafe block anyway because these functions *should* be marked `unsafe`.

Reply:
Suggestions have been applied.